### PR TITLE
Refactor tx plan computation into recursive

### PR DESCRIPTION
**Pull request by PeterBenc**
Previously:
- tx plan computation was in a really ugly for cycle
- in case wallet has for example a two utxos 2 ADA and 5 ADA, there was a 50% chance that the 2 ADA utxo gets to the tx plan computation first and if user wanted to send 1.1 ADA, 0.9 would be added to the fee (since its too small for change output) 
- if a wallet didn't have registered staking key, and had less then amount 2ADA balance, no error was shown

Changes: 
 - tx plan computation refactored into recursion
 - we try to add the change to the fee only if the plan computation with all utxos failed
 - error is shown didn't have registered staking key, and had less then amount 2ADA balance

### CODE REVIEW

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts [hunk 1]
```diff
@@ -284,59 +305,103 @@ export function computeTxPlan(
...
+    },
   }
+}
 
-  // we cant build the transaction with big enough change lovelace
-  if (change.tokenBundle.length > 0 && change.coins < minimalChangeLovelace) {
-    throw new InternalError(InternalErrorReason.ChangeOutputTooSmall)
-  }
+// TODO: remove this altogether, since shelley all utxos are profitable
+export const isUtxoProfitable = (utxo: UTxO): boolean => {
```
Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:344]: can't we simply remove this?

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts [hunk 2]
```diff
@@ -452,41 +516,24 @@ export const selectMinimalTxPlan = (
...
-            (acc, output) =>
-              output.tokenBundle.length > 0
-                ? acc + computeMinUTxOLovelaceAmount(output.tokenBundle)
-                : 0,
-            0
-          ) as Lovelace,
-        }
-      }
+  const _selectMinimalTxPlan = (
+    numOfInputs: number,
```
Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:520]: I'd name this just `numInputs`

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 1 above)
Comment (refi93) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts:344]: although technically, there can still be legacy inputs that are not profitable and it would be sensible to filter them out

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts [hunk 3]
```diff
@@ -284,59 +305,103 @@ export function computeTxPlan(
...
     return {
+      success: true,
+      txPlan: {
+        inputs,
+        outputs,
+        change: null,
+        certificates,
+        deposit,
+        additionalLovelaceAmount,
+        fee: (feeWithChange + change.coins) as Lovelace,
```
Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:322]: shouldn't this be "fee without change"?

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts [hunk 4]
```diff
@@ -452,41 +516,24 @@ export const selectMinimalTxPlan = (
...
+    allowAddingChangeToFee: boolean
+  ): TxPlanResult => {
+    const inputs: TxInput[] = utxos.slice(0, numOfInputs)
+    const txPlanResult = validateTxPlan(
+      computeTxPlan(inputs, outputs, change, certificates, withdrawals, allowAddingChangeToFee)
+    )
+    if (txPlanResult.success === true) {
+      return txPlanResult
     }
+    return numOfInputs < utxos.length
```
Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:530]: do we really need recursion? wouldn't it work as well with an ordinary while loop which is easier to follow?

Regarding the `allowAddingChangeToFee` I'd rather enhance the output of `computeTxPlan` to be able to tell how the fee was obtained, whether any change output has been "burned" - for that I'd consider just adding a `baseFee` property telling the fee attributed to the transaction as such (i.e. based on its base size) and that could be used to infer whether an output has been burned and hence whether it's worth trying to add additional inputs or not

**Review by refi93 — CHANGES_REQUESTED** [review]
see comments - I'm not completely bought into why we need recursion and I'd rather avoid it for the sake of readability/maintainability

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 1 above)
Comment (PeterBenc) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts:344]: It may seem from the diff that I somehow touched this code, but I haven't.. :smile: but yes lets remove it. Regarding the legacy utxos, yes there still can be legacy utxos that are not profitable but we agreed back in August last year that we are gonna use these utxos anyway since we are converting everything to staking balance. I think a user don't really want having a `Non-staking Balance` of 0.00001 ada.  Anyway Im gonna remove it if you don't have any objections.

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 4 above)
Comment (PeterBenc) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts:530]: I agree, I'm not completely bought into using recursion as well, I can't say I like it very much :smile: but this is rather a simple one (basically just passing an incrementing counter) so I considered it fine. 

I m gonna try the approach you proposed and if It works I am all for it. :smile:

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 3 above)
Comment (PeterBenc) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts:322]: Thnx, it seems it should :+1:

Comment (PeterBenc) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts]: Interesting, I couldnt wrap my head around how it previously worked if it is wrong, but it turned out to be correct.
`(feeWithoutChange + (change.coins + (feeWithChange - feeWithoutChange)))` which is basically same as `feeWithChange + change.coins`
We can discuss this in slack but maybe you are already getting that the change amount is firstly calculated like
`totalInput - totalOutput - feeWithChange` so for this special condition we need to add the difference between `feeWithoutChange` and `feeWithChange` to the change.

**Review by PeterBenc — COMMENTED** [review]
I fixed what you suggested, tho I need to look how is calculated fee in case there are rewards and we add change output to fee. I putting this on hold for now as I have to concentrate on other project for couple of days. I ll ping you when ist ready

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts [hunk 5]
```diff
@@ -286,57 +309,94 @@ export function computeTxPlan(
   // if we cannot create a change output with minimal ada we add it to the fee
   if (change.tokenBundle.length === 0 && change.coins < minimalChangeLovelace) {
```
Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:310]: will this work for multiasset though? You can "burn" the ADA as fees, but that's not the case with tokens :thinking:

Comment (refi93) [inline app/frontend/wallet/shelley/shelley-transaction-planner.ts:310]: I see no logic for failure in case the change output contains tokens and the ADA is not enough to cover the minimum ADA per output - how is that covered?

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 3 above)
Comment (refi93) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts]: I see, basically you need to discard the change output ADA towards the fee that was computed for the transaction containing that output - rather than putting a long expressions that is obviously simplifiable I'd put a comment explaining that - that's more valuable for somebody trying to understand what's going on

**Comment by refi93** [comment]
> I fixed what you suggested, tho I need to look how is calculated fee in case there are rewards and we add change output to fee. I putting this on hold for now as I have to concentrate on other project for couple of days. I ll ping you when ist ready

I see - that actually crossed my mind as well and conceptually, any withdrawals should count towards "totalInputs" when computing the fee

File: app/frontend/wallet/shelley/shelley-transaction-planner.ts (hunk 5 above)
Comment (PeterBenc) [reply app/frontend/wallet/shelley/shelley-transaction-planner.ts:303]: Check out the `validateTxPlan` function, it checks every output for its minimal Lovelace amount. Note that I intentionally moved all validations to the `validateTxPlan` function and left the `computeTxPlan` only the responsibility of computing a balanced plan.

**Comment by refi93** [comment]
If possible, I'd add a test case covering the edge-case outlined in the PR description, other than that, LGTM

**Review by refi93 — CHANGES_REQUESTED** [review]
If possible, I'd add a unit test covering the edge-case outlined in the PR description (as an assurance that it works correctly, otherwise hard to test manually, and to prevent possible regressions), other than that, LGTM

File: app/tests/src/transaction-planner.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    txHash: 'deadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeefdeadbeef',
+    address:
+      'addr1q8eakg39wqlye7lzyfmh900s2luc99zf7x9vs839pn4srjs2s3ps2plp2rc2qcgfmsa8kx2kk7s9s6hfq799tmcwpvpsjv0zk3',
+    coins: 2000000,
+    outputIndex: 0,
+    tokenBundle: [],
+  },
+}
+
+const successPlanSettings = {
```
Comment (refi93) [inline app/tests/src/transaction-planner.js:24]: I'd name this successPlanFixtures, which is a more standard terminology for tests

**Review by refi93 — APPROVED** [review]
LGTM, one minor naming remark
